csi/node_server.go

```diff
@@ -288,46 +288,64 @@ func (ns *NodeServer) nodeStageSharedVolume(volumeID, shareEndpoint, targetPath
 	exportPath := uri.Path
 	export := fmt.Sprintf("%s:%s", server, exportPath)
 
-	defaultMountOptions := []string{
-		"vers=4.1",
-		"noresvport",
-		//"sync",    // sync mode is prohibitively expensive on the client, so we allow for host defaults
-		//"intr",
-		//"hard",
-		//"softerr", // for this release we use soft mode, so we can always cleanup mount points
-		"timeo=600", // This is tenths of a second, so a 60 second timeout, each retrans the timeout will be linearly increased, 60s, 120s, 240s, 480s, 600s(max)
-		"retrans=5", // We try the io operation for a total of 5 times, before failing
-	}
-
-	mountOptions := append(defaultMountOptions, []string{"softerr"}...)
 	if len(customMountOptions) != 0 {
-		mountOptions = customMountOptions
-	}
+		log.Infof("Mounting shared volume %v on node %v via share endpoint %v with custom mount options %v", volumeID, ns.nodeID, shareEndpoint, customMountOptions)
+		err = mounter.Mount(export, targetPath, fsType, customMountOptions)
+	} else {
+		genDefaultMountOptions := func(nfsVer, mountMode string) []string {
+			return []string{
+				"vers=" + nfsVer,
+				"noresvport",
+				//"sync",    // sync mode is prohibitively expensive on the client, so we allow for host defaults
+				//"intr",
+				//"hard",
+				//"softerr",
+				"timeo=600", // This is tenths of a second, so a 60 second timeout, each retrans the timeout will be linearly increased, 60s, 120s, 240s, 480s, 600s(max)
+				"retrans=5", // We try the io operation for a total of 5 times, before failing
+				mountMode,
+			}
+		}
+		nfsVersToTry := []string{"4.0", "4.1"}
+		for len(nfsVersToTry) > 0 {
+			nfsVer := nfsVersToTry[0]
+			nfsVersToTry = nfsVersToTry[1:]
```

Comment:
The try & error is needed since Linux mount tool always try the highest protocol version first if there are multiple acceptable versions in the mount option.